pkg/virt-controller/watch/drain/disruptionbudget/disruptionbudget.go

```diff
@@ -314,43 +315,90 @@ func (c *DisruptionBudgetController) execute(key string) error {
 	}
 
 	// Only consider pdbs which belong to this vmi
-	pdb, err := c.pdbForVMI(namespace, name)
+	pdbs, err := c.pdbsForVMI(namespace, name)
 	if err != nil {
 		log.DefaultLogger().Reason(err).Error("Failed to fetch pod disruption budgets for namespace from cache.")
 		// If the situation does not change there is no benefit in retrying
 		return nil
 	}
 
-	return c.sync(key, vmi, pdb)
+	if len(pdbs) == 0 {
+		return c.sync(key, vmi, nil)
+	}
+	for i := range pdbs {
+		if err := c.sync(key, vmi, pdbs[i]); err != nil {
+			return err
+		}
+	}
+	return nil
+
+}
+
+func (c *DisruptionBudgetController) isMigrationComplete(vmi *virtv1.VirtualMachineInstance, migrationName string) (bool, error) {
+	notFoundErrorCheck := func(err error) (bool, error) {
+		if errors.IsNotFound(err) {
+			return true, nil
+		}
+		return false, err
+	}
+
+	vmim, err := c.clientset.VirtualMachineInstanceMigration(vmi.Namespace).Get(migrationName, &v1.GetOptions{})
+	if err != nil {
+		return notFoundErrorCheck(err)
+	}
 
+	if vmim.IsFinal() {
+		listOptions := v1.ListOptions{
+			LabelSelector: fmt.Sprintf("%v=%v", virtv1.CreatedByLabel, vmi.UID),
+			FieldSelector: fmt.Sprintf("status.phase==%v", corev1.PodRunning),
+		}
+		pods, err := c.clientset.CoreV1().Pods(vmi.Namespace).List(context.Background(), listOptions)
+		if err != nil {
+			return notFoundErrorCheck(err)
+		}
+		return len(pods.Items) < 2, nil
+	}
+	return false, nil
 }
 
 func (c *DisruptionBudgetController) sync(key string, vmi *virtv1.VirtualMachineInstance, pdb *v1beta1.PodDisruptionBudget) error {
 	delete := false
 	create := false
+
 	// delete if there is no VMI
 	if vmi == nil && pdb != nil {
 		delete = true
 	} else if vmi != nil {
 		wantsToMigrate := wantsToMigrateOnDrain(vmi)
-		if vmi.DeletionTimestamp != nil && pdb != nil {
-			// pdb can already be deleted, shutdown already in process
-			delete = true
-		} else if !wantsToMigrate && pdb != nil {
-			// We don't want migrations on evictions, if there is a pdb, remove it
-			delete = true
-		} else if wantsToMigrate && vmi.DeletionTimestamp == nil && pdb == nil {
-			// No pdb and we want migrations on evictions
-			create = true
-		} else if wantsToMigrate && pdb != nil {
-			if ownerRef := v1.GetControllerOf(pdb); ownerRef != nil && ownerRef.UID != vmi.UID {
-				// The pdb is from an old vmi with a different uid, delete and later create the correct one
-				// The VMI always has a minimum grace period, so normally this should not happen, therefore no optimizations
+
+		if pdb != nil {
```

Comment:
can we split this up into some smaller functions? I think everything is logically staying the same but it's very long and nested

Reply:
done!